

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Running Freyja on other pathogens - Measles &mdash; Freyja  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=9edc463e" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Demix output manipulation and plotting using R" href="custom-plots-with-R.html" />
    <link rel="prev" title="Custom Plotting Tutorial" href="custom_plotting_tutorial.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            Freyja
              <img src="../../_static/freyja-logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Usage:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/demix.html">freyja demix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/variants.html">freyja variants</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/barcode-build.html">freyja barcode-build</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/get-lineage-def.html">freyja get-lineage-def</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/update.html">freyja update</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/boot.html">freyja boot</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/aggregate.html">freyja aggregate</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/plot.html">freyja plot</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/dash.html">freyja dash</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/relgrowthrate.html">freyja relgrowthrate</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/extract.html">freyja extract</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/filter.html">freyja filter</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/covariants.html">freyja covariants</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/plot-covariants.html">freyja plot-covariants</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Wiki:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="command_line_workflow.html">Command Line Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="cryptic_variants.html">Cryptic Variant Detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="custom_plotting_tutorial.html">Custom Plotting Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Running Freyja on other pathogens - Measles</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#steps"><strong>STEPS</strong></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="custom-plots-with-R.html">Demix output manipulation and plotting using R</a></li>
<li class="toctree-l1"><a class="reference internal" href="lineage_barcode_extract.html">Lineage Barcode Extract</a></li>
<li class="toctree-l1"><a class="reference internal" href="read_analysis_tutorial.html">Read Analysis Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="terra_workflow.html">Terra Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="expanded_pathogens.html">Running Freyja on other pathogens</a></li>
<li class="toctree-l1"><a class="reference internal" href="custom_barcodes.html">Creating custom barcodes</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Freyja</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Running Freyja on other pathogens - Measles</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/src/wiki/freyja-measles.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="running-freyja-on-other-pathogens-measles">
<h1>Running Freyja on other pathogens - Measles<a class="headerlink" href="#running-freyja-on-other-pathogens-measles" title="Link to this heading"></a></h1>
<p>If you want to detect other pathogens using freyja,
this is a tutorial  on how to run the pipeline for Measles samples.</p>
<section id="steps">
<h2><strong>STEPS</strong><a class="headerlink" href="#steps" title="Link to this heading"></a></h2>
<p>1. Download the reference genome for your pathogen.
It can be downloaded
from NCBI database or use the references provided by <a class="reference external" href="https://nextstrain.org">Nextstrain</a> .
For the purpose of this tutorial, we will be using <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/NC_001498.1">NC_001498.1.</a>
The reference genome file can also be found <a class="reference external" href="https://github.com/andersen-lab/Freyja/blob/main/docs/data/measles-reference.fasta">here.</a></p>
<p>2. Prepare your primer file in a bed format. For this tutorial, we will be using <a class="reference external" href="https://labs.primalscheme.com/detail/artic-measles/400/v1.0.0/?q=measles">artic-measles V1.0.0 panel</a>
also provided <a class="reference external" href="https://github.com/andersen-lab/Freyja/blob/main/docs/data/artic-measles-v1.0.0.bed">here.</a>
This file will be used to simulate measles amplicon sequencing reads and trim the primers in the downstream analysis.</p>
<p>3. Prepare your reads, by assessing quality of the reads and removing the sequencing adapters.
We will be using simulated reads produced by <a class="reference external" href="https://github.com/andersen-lab/Bygul">Bygul,</a> an amplicon read simulator developed by our team.</p>
<p>For instance, we generated reads from two different measles samples using this code as following:</p>
<ul class="simple">
<li><p>Pure sample: <a class="reference external" href="https://github.com/andersen-lab/Freyja/blob/main/docs/data/GCA_031128185.1-simulated.fastq">Reads</a> simulated from a H1 lineage sample individually.</p></li>
<li><p>Mixed samples: <a class="reference external" href="https://github.com/andersen-lab/Freyja/blob/main/docs/data/measles-mixed-simulated.fastq">Reads</a> with mixing proportions H1:20% and D9:80%.</p></li>
</ul>
<p><em>Pure sample read simulation:</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bygul</span> <span class="n">simulate</span><span class="o">-</span><span class="n">proportions</span> <span class="n">GCA_031128185</span><span class="mf">.1</span><span class="o">.</span><span class="n">fna</span> <span class="n">primer</span><span class="o">.</span><span class="n">bed</span> <span class="n">measles</span><span class="o">-</span><span class="n">reference</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">outdir</span> <span class="n">measles</span><span class="o">-</span><span class="n">H1</span><span class="o">-</span><span class="mi">100</span><span class="o">/</span>
</pre></div>
</div>
<p><em>Mixed sample read simulation:</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bygul</span> <span class="n">simulate</span><span class="o">-</span><span class="n">proportions</span> <span class="n">GCA_031128185</span><span class="mf">.1</span><span class="o">.</span><span class="n">fna</span><span class="p">,</span><span class="n">GCA_031129565</span><span class="mf">.1</span><span class="o">.</span><span class="n">fna</span> <span class="n">primer</span><span class="o">.</span><span class="n">bed</span> <span class="n">measles</span><span class="o">-</span><span class="n">reference</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">outdir</span> <span class="n">measles</span><span class="o">-</span><span class="n">H1</span><span class="o">-</span><span class="mi">20</span><span class="o">-</span><span class="n">D9</span><span class="o">-</span><span class="mi">80</span><span class="o">/</span> <span class="o">--</span><span class="n">proportions</span> <span class="mf">0.2</span><span class="p">,</span><span class="mf">0.8</span>
</pre></div>
</div>
<p>4. Align your reads to your reference genome using an aligner of your choice.
Here, we use <code class="docutils literal notranslate"><span class="pre">minimap2</span></code> with parameters set for aligning short reads to a reference genome.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">minimap2</span> <span class="o">-</span><span class="n">ax</span> <span class="n">sr</span> <span class="n">reference</span><span class="o">.</span><span class="n">fasta</span> <span class="n">reads</span><span class="o">.</span><span class="n">fastq</span> <span class="o">&gt;</span> <span class="n">aligned</span><span class="o">.</span><span class="n">sam</span>
</pre></div>
</div>
<hr class="docutils" />
<p><strong>From here on, the instruction will be the same for mixed and pure samples.
Please change the file names accordingly.</strong></p>
<ol class="arabic simple" start="5">
<li><p>Convert sam format file to a bam file using samtools.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">bS</span> <span class="n">aligned</span><span class="o">.</span><span class="n">sam</span> <span class="o">&gt;</span> <span class="n">aligned</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<ol class="arabic simple" start="6">
<li><p>Sort your bam file.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">sort</span> <span class="o">-</span><span class="n">o</span> <span class="n">aligned_sorted</span><span class="o">.</span><span class="n">bam</span> <span class="n">aligned</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<ol class="arabic simple" start="7">
<li><p>Index your bam file.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">index</span> <span class="n">aligned_sorted</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>8. Remove primers from the reads. The following command will remove reads with mean
quality score <cite>-q</cite> less than 15 and minimum read length of 50 (after trimming).
Flag <cite>-e</cite> keeps reads without primers in the output and
the flag <cite>-x</cite> will make sure that reads that occur at the
specified offset positions relative to primer positions will also be trimmed.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ivar</span> <span class="n">trim</span> <span class="o">-</span><span class="n">b</span> <span class="n">primer</span><span class="o">.</span><span class="n">bed</span> <span class="o">-</span><span class="n">p</span> <span class="n">trimmed</span> <span class="o">-</span><span class="n">i</span> <span class="n">aligned_sorted</span><span class="o">.</span><span class="n">bam</span> <span class="o">-</span><span class="n">q</span> <span class="mi">15</span> <span class="o">-</span><span class="n">m</span> <span class="mi">50</span> <span class="o">-</span><span class="n">e</span> <span class="o">-</span><span class="n">x</span> <span class="mi">3</span>
</pre></div>
</div>
<ol class="arabic simple" start="9">
<li><p>Sort and index the trimmed bam file.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">sort</span> <span class="o">-</span><span class="n">o</span> <span class="n">trimmed_sorted</span><span class="o">.</span><span class="n">bam</span> <span class="n">trimmed</span><span class="o">.</span><span class="n">bam</span> <span class="o">&amp;&amp;</span> <span class="n">samtools</span> <span class="n">index</span> <span class="n">trimmed_sorted</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>10. Generate coverage depth for each single genomic location in the reference.
This will be used in freyja pipeline downstream analysis.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">mpileup</span> <span class="o">-</span><span class="n">aa</span> <span class="o">-</span><span class="n">A</span> <span class="o">-</span><span class="n">d</span> <span class="mi">600000</span> <span class="o">-</span><span class="n">Q</span> <span class="mi">20</span> <span class="o">-</span><span class="n">q</span> <span class="mi">0</span> <span class="o">-</span><span class="n">B</span> <span class="o">-</span><span class="n">f</span> <span class="n">reference</span><span class="o">.</span><span class="n">fasta</span> <span class="n">trimmed_sorted</span><span class="o">.</span><span class="n">bam</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f1</span><span class="o">-</span><span class="mi">4</span> <span class="o">&gt;</span> <span class="n">depths</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
<ol class="arabic simple" start="11">
<li><p>Call variants using variant caller of your choice. We recommend using Lofreq or ivar and included both commands for your reference.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># call variants using lofreq</span>
<span class="n">lofreq</span> <span class="n">call</span> <span class="o">-</span><span class="n">f</span> <span class="n">reference</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">a</span> <span class="mi">1</span> <span class="o">-</span><span class="n">b</span> <span class="mi">1</span> <span class="o">-</span><span class="n">o</span> <span class="n">variants</span><span class="o">.</span><span class="n">vcf</span> <span class="n">trimmed_sorted</span><span class="o">.</span><span class="n">bam</span>
<span class="c1"># Call variants using ivar</span>
<span class="n">freyja</span> <span class="n">variants</span> <span class="n">trimmed_sorted</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">variants</span> <span class="n">variants</span><span class="o">.</span><span class="n">tsv</span> <span class="o">--</span><span class="n">depths</span> <span class="n">depths</span><span class="o">.</span><span class="n">tsv</span> <span class="o">--</span><span class="n">ref</span> <span class="n">reference</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<ol class="arabic simple" start="12">
<li><p>Run freyja demix to estimate lineage prevalence.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">freyja</span> <span class="n">demix</span> <span class="n">variants</span><span class="o">.</span><span class="n">tsv</span> <span class="n">depths</span><span class="o">.</span><span class="n">tsv</span> <span class="o">--</span><span class="n">output</span> <span class="n">freyja_demix</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">pathogen</span> <span class="n">MEASLESgenome</span>
</pre></div>
</div>
<p>The final demix outputs for the pure and mixed sample are as following:</p>
<p><em>Mixed sample output:</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">summarized</span>      <span class="p">[(</span><span class="s1">&#39;Other&#39;</span><span class="p">,</span> <span class="mf">0.9999999968413253</span><span class="p">)]</span>
<span class="n">lineages</span>        <span class="n">MEASLES</span><span class="o">-</span><span class="n">D9</span> <span class="n">MEASLES</span><span class="o">-</span><span class="n">H1</span>
<span class="n">abundances</span>      <span class="mf">0.79692605</span> <span class="mf">0.20307394</span>
<span class="n">resid</span>   <span class="mf">214.51679168207156</span>
<span class="n">coverage</span>        <span class="mf">91.39927016484208</span>
</pre></div>
</div>
<p><em>Pure sample output</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">summarized</span>      <span class="p">[(</span><span class="s1">&#39;Other&#39;</span><span class="p">,</span> <span class="mf">0.999999999926792</span><span class="p">)]</span>
<span class="n">lineages</span>        <span class="n">MEASLES</span><span class="o">-</span><span class="n">H1</span>
<span class="n">abundances</span>      <span class="mf">1.00000000</span>
<span class="n">resid</span>   <span class="mf">53.868769540487826</span>
<span class="n">coverage</span>        <span class="mf">89.52434881087203</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="custom_plotting_tutorial.html" class="btn btn-neutral float-left" title="Custom Plotting Tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="custom-plots-with-R.html" class="btn btn-neutral float-right" title="Demix output manipulation and plotting using R" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Andersen Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>