<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Lineage Barcode Extract &mdash; Freyja 1.4.9 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=1699af0b"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Read Analysis Tutorial" href="read_analysis_tutorial.html" />
    <link rel="prev" title="Custom Plotting Tutorial" href="custom_plotting_tutorial.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Freyja
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Freyja:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Wiki:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="command_line_workflow.html">Command Line Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="cryptic_variants.html">Cryptic Variant Detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="custom_plotting_tutorial.html">Custom Plotting Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Lineage Barcode Extract</a></li>
<li class="toctree-l1"><a class="reference internal" href="read_analysis_tutorial.html">Read Analysis Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="terra_workflow.html">Terra Workflow</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Freyja</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Lineage Barcode Extract</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/lineage_barcode_extract.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="lineage-barcode-extract">
<h1>Lineage Barcode Extract<a class="headerlink" href="#lineage-barcode-extract" title="Link to this heading"></a></h1>
<p>If you’re looking to identify lineage defining mutations for your lineage of choice, look no further! You can extract this information using Freyja’s existing barcoding and a bit of python.</p>
<ol class="arabic simple">
<li><p>Run a freyja update and extract barcodes to the directory of your choosing (will take a few minutes):</p></li>
</ol>
<p><code class="docutils literal notranslate"><span class="pre">`</span>
<span class="pre">freyja</span> <span class="pre">update</span> <span class="pre">--outdir</span> <span class="pre">/my/local/directory/</span>
<span class="pre">`</span></code></p>
<ol class="arabic simple" start="2">
<li><p>Load the barcodes using pandas and extract barcodes for your lineages of interest</p></li>
</ol>
<p><a href="#id1"><span class="problematic" id="id2">``</span></a><a href="#id3"><span class="problematic" id="id4">`</span></a>python
import pandas as pd</p>
<dl class="simple">
<dt>def sortFun(x):</dt><dd><p># sort based on nuc position, ignoring nuc identities
return int(x[1:(len(x)-1)])</p>
</dd>
</dl>
<p># replace “/my/local/directory/” with your chosen directory
df = pd.read_csv(‘/my/local/directory/usher_barcodes.csv’, index_col=0)</p>
<p># specify your lineages of interest here
lins = [‘B.1.617.2’,’BA.1’,’BA.2’,’B.1.1.7’,’P.1’]
df = df.loc[lins]
# keep only columns with at least one mutation across the lins
keepcols = df.columns[df.sum(axis=0)&gt;0]
df = df.loc[:, keepcols]</p>
<p><a href="#id5"><span class="problematic" id="id6">``</span></a>`
Optional:
3. Load in gene coordinate information to translate nucleotide mutations into possible AA mutations. An example file for doing this is available [here](<a class="reference external" href="https://github.com/andersen-lab/Freyja/wiki/SARS-CoV-2.json">https://github.com/andersen-lab/Freyja/wiki/SARS-CoV-2.json</a>). Then append AA mutations to the corresponding nucleotide mutations.</p>
<p><a href="#id7"><span class="problematic" id="id8">``</span></a><a href="#id9"><span class="problematic" id="id10">`</span></a>python
# build dataframe with gene data
import json
f1 = open(‘/my/local/directory/SARS-CoV-2.json’,)
dat = json.load(f1)
f1.close()
df_genes = pd.DataFrame(columns=[“gene”,”start”,”end”], dtype=object)</p>
<dl>
<dt>for d in dat[“genes”]:</dt><dd><dl class="simple">
<dt>new_row = {‘gene’:d,’start’:dat[“genes”][d][‘coordinates’][‘from’],</dt><dd><p>‘end’:dat[“genes”][d][‘coordinates’][‘to’]}</p>
</dd>
</dl>
<p>df_genes = df_genes.append(new_row,ignore_index=True)</p>
</dd>
</dl>
<p>df_genes = df_genes.sort_values(‘start’).set_index(‘gene’,drop=False)</p>
<p># function for converting to gene-wise coordinate numbering
def getGeneNum(pos,df0):</p>
<blockquote>
<div><p>j=0
while j&lt;df0.shape[0]:</p>
<blockquote>
<div><dl class="simple">
<dt>if pos&lt;=df0.iloc[j].loc[‘end’] and pos&gt;=df0.iloc[j].loc[‘start’]:</dt><dd><p>return df0.iloc[j].loc[‘gene’], (pos - df0.iloc[j].loc[‘start’])//3+1</p>
</dd>
<dt>else:</dt><dd><p>j+=1</p>
</dd>
</dl>
</div></blockquote>
<p>return ‘’,’’</p>
</div></blockquote>
<p># add AA mutation info into our dataframe
cols = list(df.columns)
cols.sort(key=lambda x: sortFun(x))
df = df[cols]
for i, mut in enumerate(cols):</p>
<blockquote>
<div><p>pos = sortFun(mut)
g,pa = getGeneNum(pos,df_genes)
if len(g)==0:</p>
<blockquote>
<div><p>refAA=’’
mutAA=’’</p>
</div></blockquote>
<dl class="simple">
<dt>else:</dt><dd><p>posInTriple = (pos - df_genes.loc[g,’start’]) % 3
triple = ref0[pos-posInTriple-1:pos-posInTriple+2]
refAA = triple.translate()
tripleMut = triple
tripleMut[posInTriple] = mut[-1]
mutAA = tripleMut.translate()
cols[i] = cols[i] +’(’+ g+’:’+str(refAA)+str(pa)+str(mutAA)+’)’</p>
</dd>
</dl>
</div></blockquote>
<p>df.columns = cols
<a href="#id11"><span class="problematic" id="id12">``</span></a><a href="#id13"><span class="problematic" id="id14">`</span></a></p>
<p>4. Export to csv format for later viewing.
<code class="docutils literal notranslate"><span class="pre">`python</span>
<span class="pre">df.to_csv('mutations_lineage_subset.csv')</span>
<span class="pre">`</span></code></p>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="custom_plotting_tutorial.html" class="btn btn-neutral float-left" title="Custom Plotting Tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="read_analysis_tutorial.html" class="btn btn-neutral float-right" title="Read Analysis Tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Andersen Lab development team.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>